File: OC/TradePlatform/Class/Home(首页)/View/ServiceModuleView.m
//
//  ServiceModuleView.m
//  TradePlatform
//
//  Created by apple on 2017/5/17.
//  Copyright © 2017年 apple. All rights reserved.
//

#import "ServiceModuleView.h"

@interface ServiceModuleView ()<UICollectionViewDelegate, UICollectionViewDataSource>

/** collection样式 */
@property (strong, nonatomic) UICollectionViewFlowLayout *collectionViewFlowLayout;
/** collection */
@property (strong, nonatomic) UICollectionView *collectionView;
/** pageControl */
@property (nonatomic, strong) SMPageControl *pageControl;

@end

@implementation ServiceModuleView

- (instancetype)init {
    self = [super init];
    if (self) {
        self.backgroundColor = WhiteColor;
        [self serviceModuleLayoutView];
    }
    return self;
}

#pragma mark - 获取服务模块数据set方法
- (void)setModuleArray:(NSMutableArray *)moduleArray {
    _moduleArray = moduleArray;
    self.pageControl.numberOfPages = moduleArray.count;
    [self.collectionView reloadData];
}


#pragma mark - 服务模块点击方法
- (void)moduleBtnAction:(UIButton *)button {
    if (_delegate && [_delegate respondsToSelector:@selector(serviceModuleBtnAction:)]) {
        [_delegate serviceModuleBtnAction:button];
    }
}


#pragma mark - collection代理方法
// 告诉系统一共多少组
- (NSInteger)numberOfSectionsInCollectionView:(UICollectionView *)collectionView {
    return self.moduleArray.count;
}
// 告诉系统每组多少个
- (NSInteger)collectionView:(UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    NSArray *moduleItem = [self.moduleArray objectAtIndex:section];
    return moduleItem.count;
}
// 告诉系统每个Cell如何显示
- (UICollectionViewCell *)collectionView:(UICollectionView *)collectionView cellForItemAtIndexPath:(NSIndexPath *)indexPath {
    ModuleCell *cell = [collectionView dequeueReusableCellWithReuseIdentifier:@"moduleCell" forIndexPath:indexPath];
    NSArray *moduleItem = [self.moduleArray objectAtIndex:indexPath.section];
    NSArray *itemArray = [moduleItem objectAtIndex:indexPath.row];
    /** cell位置 */
    cell.indexPath = indexPath;
    /** 按钮数据 */
    cell.btnArray = itemArray;
    /** 按钮1 */
    [cell.btnOne.moduleBtn addTarget:self action:@selector(moduleBtnAction:) forControlEvents:UIControlEventTouchUpInside];
    /** 按钮2 */
    [cell.btnTwo.moduleBtn addTarget:self action:@selector(moduleBtnAction:) forControlEvents:UIControlEventTouchUpInside];
    /** 按钮3 */
    [cell.btnThree.moduleBtn addTarget:self action:@selector(moduleBtnAction:) forControlEvents:UIControlEventTouchUpInside];
    return cell;
}
// 定义每个Cell的大小
-(CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout *)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath {
    return CGSizeMake(ScreenW - 32, 106);;
}
// 这个是两行cell之间的间距（上下行cell的间距）
- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout minimumLineSpacingForSectionAtIndex:(NSInteger)section {
    return 0;
}
// 两个cell之间的间距（同一行的cell的间距）
- (CGFloat)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout minimumInteritemSpacingForSectionAtIndex:(NSInteger)section {
    return 0;
}
// 动态设置某组头视图大小
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout referenceSizeForHeaderInSection:(NSInteger)section {
    return CGSizeMake(16, 0);
}


//动态设置某组尾视图大小
- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout referenceSizeForFooterInSection:(NSInteger)section {
    return CGSizeMake(16, 0);
}


#pragma mark - scrollView代理方法
- (void)scrollViewDidScroll:(UIScrollView *)scrollView {
    int page = scrollView.contentOffset.x / scrollView.width + 0.5;
    self.pageControl.currentPage = page;
}


#pragma mark - view布局
- (void)serviceModuleLayoutView {
    /** collection样式 */
    self.collectionViewFlowLayout = [[UICollectionViewFlowLayout alloc] init];
    // 设置滚动方向
    self.collectionViewFlowLayout.scrollDirection = UICollectionViewScrollDirectionHorizontal;
    /** collection */
    self.collectionView = [[UICollectionView alloc] initWithFrame:self.frame collectionViewLayout:self.collectionViewFlowLayout];
    self.collectionView.backgroundColor = CLEARCOLOR;
    self.collectionView.delegate = self;
    self.collectionView.dataSource = self;
    self.collectionView.showsHorizontalScrollIndicator = NO;
    self.collectionView.pagingEnabled = YES;
    [self addSubview:self.collectionView];
    // 注册Item
    [self.collectionView registerClass:[ModuleCell class] forCellWithReuseIdentifier:@"moduleCell"];
    // 添加pageControl
    self.pageControl = [[SMPageControl alloc] init];
    self.pageControl.userInteractionEnabled = NO;
    self.pageControl.indicatorMargin = 10.0f;
    self.pageControl.indicatorDiameter = 10.0f;
    self.pageControl.pageIndicatorImage = [UIImage imageNamed:@"home_page_service_unchecked"];
    self.pageControl.currentPageIndicatorImage = [UIImage imageNamed:@"home_page_service_checked"];
    [self addSubview:_pageControl];
}

- (void)layoutSubviews {
    [super layoutSubviews];
    @weakify(self)
    /** collection */
    [self.collectionView mas_makeConstraints:^(MASConstraintMaker *make) {
        @strongify(self)
        make.top.equalTo(self.mas_top);
        make.bottom.equalTo(self.mas_bottom).offset(-20);
        make.left.equalTo(self.mas_left);
        make.right.equalTo(self.mas_right);
    }];
    /** pageControl */
    [self.pageControl mas_makeConstraints:^(MASConstraintMaker *make) {
        make.top.equalTo(self.collectionView.mas_bottom);
        make.centerX.equalTo(self.mas_centerX);
        make.height.mas_equalTo(@20);
        make.width.mas_equalTo(@100);
    }];
    /** self */
    [self mas_makeConstraints:^(MASConstraintMaker *make) {
        make.height.mas_equalTo(@232);
    }];
}

@end

File: Swict/LoveCarArtisan/Class/Home(首页)/View/ServiceModuleView.swift
//
//  ServiceModuleView.swift
//  LoveCarArtisan
//
//  Created by apple on 2017/5/18.
//  Copyright © 2017年 apple. All rights reserved.
//

import UIKit

// MARK: 定义代理
protocol ServiceModuleDelegate {
    //代理方法
    func moduleBtnDelegate(button: UIButton)
}

class ServiceModuleView: UIView {
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    override init(frame: CGRect) {
        super.init(frame: frame)
        serviceModulelaoutView()
    }
    // MARK: 定义代理
    var delegate : ServiceModuleDelegate?
    
    // MARK: 定义控件
    /** 服务模块数据 */
    var moduleArray: NSMutableArray = NSMutableArray() {
        didSet {
            pageControl.numberOfPages = moduleArray.count
            collectionView?.reloadData()
        }
    }
    /** collection样式 */
    private var flowLayout: UICollectionViewFlowLayout = UICollectionViewFlowLayout()
    /** collection */
    private var collectionView: UICollectionView?
    /** pageControl */
    lazy var pageControl: UIPageControl = {
        let pageControl = UIPageControl()
        pageControl.currentPage = 0
        pageControl.pageIndicatorTintColor = GrayH6Color
        pageControl.currentPageIndicatorTintColor = ThemeColor
        return pageControl
    }()
    // MARK: 布局控件
    func serviceModulelaoutView() {
        /** collection样式 */
        flowLayout = UICollectionViewFlowLayout()
        // 设置滚动方向
        flowLayout.scrollDirection = UICollectionViewScrollDirection.horizontal;
        /** collection */
        collectionView = UICollectionView(frame: self.frame, collectionViewLayout: flowLayout)
        collectionView!.backgroundColor = UIColor.clear;
        collectionView!.delegate = self;
        collectionView!.dataSource = self;
        collectionView!.showsHorizontalScrollIndicator = false;
        collectionView!.isPagingEnabled = true;
        // 注册Item
        collectionView!.register(ModuleCell.self, forCellWithReuseIdentifier:"moduleCell")
        addSubview(collectionView!)
        /** pageControl */
        addSubview(pageControl)
    }
    override func layoutSubviews() {
        super.layoutSubviews()
        /** collection */
        collectionView!.mas_makeConstraints { (make:MASConstraintMaker!) in
            make.left.equalTo()(self.mas_left);
            make.right.equalTo()(self.mas_right);
            make.top.equalTo()(self.mas_top);
            make.bottom.equalTo()(self.mas_bottom)?.setOffset(-20);
        }
        // 页面控制器
        pageControl.mas_makeConstraints { (make:MASConstraintMaker!) in
            make.centerX.equalTo()(self.collectionView!.mas_centerX)
            make.top.equalTo()(self.collectionView!.mas_bottom)
            make.width.equalTo()(100)
            make.height.equalTo()(20)
        }
        /** self高度 */
        mas_makeConstraints { (make:MASConstraintMaker!) in
            make.height.mas_equalTo()(232);
        }
    }
}


extension ServiceModuleView : UICollectionViewDelegate, UICollectionViewDataSource, UICollectionViewDelegateFlowLayout {
    // 告诉系统一共多少组
    func numberOfSections(in collectionView: UICollectionView) -> Int {
        return moduleArray.count
    }
    // 告诉系统每组多少个
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        let moduleItem: NSArray = moduleArray.object(at: section) as! NSArray
        return moduleItem.count
    }
    // 告诉系统每个Cell如何显示
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let cell: ModuleCell = collectionView.dequeueReusableCell(withReuseIdentifier: "moduleCell", for: indexPath) as! ModuleCell
        let moduleItem: NSArray = moduleArray.object(at: indexPath.section) as! NSArray
        let itemArray: NSArray = moduleItem.object(at: indexPath.row) as! NSArray
        /** cell位置 */
        cell.indexPath = indexPath as NSIndexPath;
        /** 按钮数据 */
        cell.btnArray = itemArray;
        /** 按钮1 */
        cell.btnOne.serviceBtn.addTarget(self, action: #selector(moduleBtnAction(button:)), for: UIControlEvents.touchUpInside)
        /** 按钮2 */
        cell.btnTwo.serviceBtn.addTarget(self, action: #selector(moduleBtnAction(button:)), for: UIControlEvents.touchUpInside)
        /** 按钮3 */
        cell.btnThree.serviceBtn.addTarget(self, action: #selector(moduleBtnAction(button:)), for: UIControlEvents.touchUpInside)
        return cell
    }
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
        return CGSize(width: AppWidth - 32, height: 106)
    }
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumLineSpacingForSectionAt section: Int) -> CGFloat {
        return 0
    }
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, minimumInteritemSpacingForSectionAt section: Int) -> CGFloat {
        return 0
    }
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForHeaderInSection section: Int) -> CGSize {
        return CGSize(width: 16, height: 0)
    }
    func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, referenceSizeForFooterInSection section: Int) -> CGSize {
        return CGSize(width: 16, height: 0)
    }
    // MARK: 按钮点击方法
    func moduleBtnAction(button: UIButton) {
        self.delegate?.moduleBtnDelegate(button: button)
    }
}


extension ServiceModuleView : UIScrollViewDelegate {
    func scrollViewDidScroll(_ scrollView: UIScrollView) {
        pageControl.currentPage = Int(scrollView.contentOffset.x / AppWidth)
    }
}




